subsys/net/lib/nrf_cloud/src/nrf_cloud_download.c

```diff
@@ -9,17 +9,271 @@
+		return dl->dlc->callback(evt);
+	}
+
+	return -EINVAL;
+}
+
+static int coap_dl_disconnect(void)
+{
+	return nrf_cloud_coap_transport_disconnect(&coap_client);
+}
+
+static void coap_cb(int16_t result_code, size_t offset, const uint8_t *payload, size_t len,
+		    bool last_block, void *user_data)
+{
+	int ret;
+	struct nrf_cloud_download_data *dl = (struct nrf_cloud_download_data *)user_data;
+	struct download_client_evt evt = {0};
+
+	LOG_DBG("CoAP result: %d, offset: 0x%X, len: 0x%X, last_block: %d",
+		result_code, offset, len, last_block);
+
+	/* TODO: save result code?? */
+
+	if ((result_code != COAP_RESPONSE_CODE_OK) && (result_code != COAP_RESPONSE_CODE_CONTENT)) {
+		LOG_ERR("Unexpected response: %*s", len, payload);
+		evt.id = DOWNLOAD_CLIENT_EVT_ERROR;
+		evt.error = ETIMEDOUT;
+		/* TODO: any additional handling? */
+
+	} else {
+		evt.id = DOWNLOAD_CLIENT_EVT_FRAGMENT;
+		evt.fragment.buf = payload;
+		evt.fragment.len = len;
+	}
+
+	ret = coap_dl_event_send(dl, &evt);
+	if (ret) {
+		/* ERROR TODO */
+	}
+
+	if (last_block) {
+		LOG_INF("Download complete");
+
+		memset(&evt, 0, sizeof(evt));
+		evt.id = DOWNLOAD_CLIENT_EVT_DONE;
+
+		ret = coap_dl_event_send(dl, &evt);
+		if (ret) {
+			/* ERROR TODO */
+		}
+
+		ret = coap_dl_disconnect();
+		if (ret) {
+			/* ERROR TODO */
+			LOG_ERR("Failed to disconnect CoAP transport, error: %d", ret);
+		}
+
+		if (dl->type == NRF_CLOUD_DL_TYPE_FOTA) {
+			/* fota_download expects a closed event */
+			memset(&evt, 0, sizeof(evt));
+			evt.id = DOWNLOAD_CLIENT_EVT_CLOSED;
+
+			ret = coap_dl_event_send(dl, &evt);
+			if (ret) {
+				/* ERROR TODO */
+			}
+		}
+	}
+
+	if (last_block || (result_code >= COAP_RESPONSE_CODE_BAD_REQUEST)) {
+		/* TODO: any error handling or completion signaling? */
+	}
+}
+
+#define MAX_RETRIES 5
+static int coap_dl_start(struct nrf_cloud_download_data *const dl)
+{
+	int err;
+	int retry = 0;
+	struct coap_client *cc = &coap_client.cc;
+
+	struct coap_client_request request = {
+		.method = COAP_METHOD_GET,
+		.confirmable = true,
+		.path = NRF_CLOUD_COAP_PROXY_RSC,
+		.fmt = COAP_CONTENT_FORMAT_APP_OCTET_STREAM,
+		.payload = NULL,
+		.len = 0,
+		.cb = coap_cb,
+		.user_data = dl,
+		.options = cc_opts,
+		.num_options = OPT_CNT
+	};
+
+	while ((err = coap_client_req(cc, cc->fd, NULL, &request, NULL)) == -EAGAIN) {
+		if (retry++ > MAX_RETRIES) {
+			LOG_ERR("Timeout waiting for CoAP client to be available");
+			return -ETIMEDOUT;
+		}
+		LOG_DBG("CoAP client busy");
+		k_sleep(K_MSEC(500));
+	}
+
+	if (err == 0) {
+		LOG_DBG("Sent CoAP download request");
+	} else {
+		LOG_ERR("Error sending CoAP request: %d", err);
+	}
+
+	return err;
+}
+
+static int coap_dl(struct nrf_cloud_download_data *const dl)
+{
+	int ret = coap_dl_init();
+
+	if (ret) {
+		/* ERROR TODO */
+		return ret;
+	}
+
+	ret = coap_dl_connect_and_auth(dl);
+	if (ret) {
+		/* ERROR TODO */
+		return ret;
+	}
+
+	/* Get the options for the proxy download */
```

Comment:
TODO: 
Need to add a call to fota_download_b1_file_parse before getting the options.